src/data/loaders.py

```diff
@@ -1,113 +1,423 @@
                 row = line.rstrip('\n').split(',', 1)
                 word_id = int(row[0])
-                self.word_ids[row[1]] = word_id
+                self.word_to_id[row[1]] = word_id
+                self.id_to_word[word_id] = row[1]
                 if word_id > self.highest_word_id:
                     self.highest_word_id = word_id
 
-        if persist_file_name is not None:
-            self.persist_file = open(persist_file_name, 'w')
-        else:
-            self.persist_file = None
+    def is_song(self, filepath):
+        return filepath.endswith('.txt')
+
+    def read(self, filepath):
+        """Read a file.
+
+        Arguments:
+            filepath (str): path to the lyrics file. e.g.
+                "/home/user/lyrics_data/tool/lateralus.txt"
+        """
+        return ''.join(open(filepath, 'r', errors='ignore').readlines())
 
-    def __call__(self, filepath):
-        """This method takes some lyrics data and returns a list of integers
-        word IDs for that lyrics data.
+    def tokenize(self, raw_lyrics):
+        """Turns a string of lyrics data into a numpy array of int "word" IDs.
 
         Arguments:
-            filepath (str): specifies the path to the file to load.
+            raw_lyrics (str): Stringified lyrics data
         """
-        tokens = self.tokenize(filepath, self.length)
-        word_ids = np.zeros(self.length, dtype=self.dtype)
-        for token_index, token in enumerate(tokens):
-            if token not in self.word_ids:
+        tokens = []
+        for token in self.tokenizer(raw_lyrics):
+            if token not in self.word_to_id:
                 self.highest_word_id += 1
-                self.word_ids[token] = self.highest_word_id
-                if self.persist_file is not None:
-                    self.persist_file.write(
-                        '%s,%s\n' % (self.highest_word_id, token))
-            word_ids[token_index] = self.word_ids[token]
-        return word_ids
+                self.word_to_id[token] = self.highest_word_id
+                self.id_to_word[self.highest_word_id] = token
+                if self.persist:
+                    self.metadata.write(
+                        'word_ids.csv',
+                        '%s,%s\n' % (self.highest_word_id, token)
+                    )
+            tokens.append(self.word_to_id[token])
+        return tokens
 
-    def close(self):
-        self.persist_file.close()
+    def detokenize(self, numpy_data):
+        ret = ''
+        for token in numpy_data:
+            word = self.id_to_word[token]
+            if word == "n't":
+                ret += word
+            elif word not in string.punctuation and not word.startswith("'"):
+                ret += " " + word
+            else:
+                ret += word
+        return "".join(ret).strip()
 
 
-class MIDILoader(object):
-    """A class which loads and parses MIDI files into a list of event IDs.
+def resolve_pitch_clashes(sorted_notes):
```

Comment:
@sagelywizard @siddharth-agrawal could we add a small description of what clash resolution scheme the code is implementing?

Reply:
Sure